[1/2] Add Kernel support for Cutlass based Fused FP4 MoE

pavanimajety opened the pull request:
This kernel adds support for NVFP4 MoE kernels. 

Currently measured perf: 
```
[--------------------------------------------------------------------------------------------- FP4 MOE vs FP8 Triton ---------------------------------------------------------------------------------------------]
                                                                                                                       |  triton_moe  |  triton_moe_cuda_graphs  |  cutlass_moe_fp4  |  cutlass_moe_fp4_cuda_graphs
1 threads: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((4, 2048, 7168))     |     10.7     |            9.9           |        12.3       |              10.5
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((8, 2048, 7168))     |     17.5     |           16.7           |        16.4       |              14.5
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((16, 2048, 7168))    |     29.5     |           28.8           |        23.4       |              21.5
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((32, 2048, 7168))    |     45.8     |           45.1           |        32.8       |              31.2
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((64, 2048, 7168))    |     60.9     |           60.0           |        41.5       |              40.3
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((128, 2048, 7168))   |     69.5     |           68.8           |        47.8       |              45.3
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((256, 2048, 7168))   |     72.3     |           71.4           |        50.9       |              49.9
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((512, 2048, 7168))   |     80.9     |           80.1           |        59.6       |              57.8
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((1024, 2048, 7168))  |     93.7     |           92.8           |        75.9       |              74.7
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((2048, 2048, 7168))  |    142.3     |          142.0           |       109.2       |             107.6

Times are in milliseconds (ms).
```

Follow up PR would include model support for `nvidia/DeepSeekR1-FP4`


<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

## Modifications

<!-- Describe the changes made in this PR. -->

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [x] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

pavanimajety committed:
Initial FP4 MoE commit

Signed-off-by: Pavani Majety <pmajety@nvidia.com>

pavanimajety committed:
remove get_sm_version

Signed-off-by: Pavani Majety <pmajety@nvidia.com>

pavanimajety committed:
Running changes

pavanimajety committed:
add test and other necessary changes in CMakeLists.txt

Signed-off-by: Pavani Majety <pmajety@nvidia.com>

pavanimajety committed:
fix pre-commit

Signed-off-by: Pavani Majety <pmajety@nvidia.com>

Fridge003 commented:
Hi @pavanimajety, please fix the lint by running `make format` under `sgl-kernel` folder

pavanimajety committed:
Combine blockscale offsets to the prepare function to reduce overhead

Signed-off-by: Pavani Majety <pmajety@nvidia.com>

pavanimajety committed:
more formatting changes and change flaky test

pavanimajety committed:
fix some warnings

pavanimajety committed:
apply format diff and fix constexpr warning

pavanimajety committed:
lint again

elfiegg commented on python/sglang/srt/layers/moe/cutlass_moe.py:
```diff
@@ -205,3 +204,153 @@ def cutlass_fused_experts(
+    out_dtype = a.dtype
+    num_topk = topk_ids.shape[1]
+
+    expert_offsets = torch.empty((e + 1), dtype=torch.int32, device=device)
```
For fp8 bs these tensors are initiated after weights are loaded and are passed into the class for perf reasons. I suggest for fp4 we can stick to that

elfiegg reviewed (commented):
Thanks Pavani for the work! I left some comments and realized it might require some structure changes and hopefully they make sense to you.

pavanimajety commented on sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu:
```diff
@@ -0,0 +1,446 @@
+  int num_experts = static_cast<int>(expert_offsets.size(0));
+  auto options_int = torch::TensorOptions().dtype(torch::kInt64).device(a.device());
+
+  torch::Tensor a_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor b_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor out_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor a_scales_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor b_scales_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor alpha_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor layout_sfa = torch::empty({num_experts, 5}, options_int);
+  torch::Tensor layout_sfb = torch::empty({num_experts, 5}, options_int);
+  torch::Tensor c_strides1 = torch::full({num_experts}, output.stride(0), options_int);
+  torch::Tensor a_strides1 = torch::full({num_experts}, a.stride(0) * 2, options_int);
+  torch::Tensor b_strides1 = torch::full({num_experts}, b.stride(1) * 2, options_int);
```
yes, this has been tested with graph mode.

pavanimajety committed:
Add ab_strides and c_strides to the kernel and add shuffle_rows kernel

Signed-off-by: Pavani Majety <pmajety@nvidia.com>

pavanimajety committed:
Fix shuffle_rows issue and lint

Signed-off-by: Pavani Majety <pmajety@nvidia.com>

pavanimajety committed:
Fix lint again

Signed-off-by: Pavani Majety <pmajety@nvidia.com>

pavanimajety committed:
Merge branch 'main' into cutlass_fp4_moe

pavanimajety commented on python/sglang/srt/layers/moe/cutlass_moe.py:
```diff
@@ -205,3 +204,153 @@ def cutlass_fused_experts(
+
+    # problem shapes should have [m, n, k]
+    # Note that problem sizes are based on logical number of elements.
+    blockscale_offsets = torch.zeros(e + 1, dtype=torch.int32, device=device)
```
Changed the logic, so initializing only empty which should have no overhead

pavanimajety commented on sgl-kernel/csrc/moe/prepare_moe_input.cu:
```diff
@@ -126,3 +146,68 @@ void prepare_moe_input(
+      topk_ids.size(1));
+}
+
+void prepare_moe_input_v2(
```
Done, passing it as an optional argument

pavanimajety commented:
@zhyncs @elfiegg I addressed the feedback. The failing tests seem unrelated to the PR. Please let me know if you have more suggestions or if it is ready for merge. Thanks!

elfiegg commented on sgl-kernel/csrc/moe/nvfp4_blockwise_moe.cu:
```diff
@@ -0,0 +1,471 @@
+  int num_experts = static_cast<int>(expert_offsets.size(0));
+  auto options_int = torch::TensorOptions().dtype(torch::kInt64).device(a.device());
+
+  torch::Tensor a_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor b_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor out_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor a_scales_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor b_scales_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor alpha_ptrs = torch::empty(num_experts, options_int);
+  torch::Tensor layout_sfa = torch::empty({num_experts, 5}, options_int);
+  torch::Tensor layout_sfb = torch::empty({num_experts, 5}, options_int);
```
Can we try moving all these static tensors out of the kernel?

zhyncs replied:
agree

pavanimajety replied:
They also go out of scope after the kernel finishes running, which also makes it slightly better I think.

zhyncs committed:
Merge branch 'main' into cutlass_fp4_moe

pavanimajety commented:
```
[--------------------------------------------------------------------------- NVFP4 Blockscaled CUTLASS MOE vs FP8 Tensor Scaled Triton ---------------------------------------------------------------------------]
                                                                                                                       |  triton_moe  |  triton_moe_cuda_graphs  |  cutlass_moe_fp4  |  cutlass_moe_fp4_cuda_graphs
1 threads: --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((4, 2048, 7168))     |      9.6     |            8.8           |         9.3       |               8.4           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((8, 2048, 7168))     |     15.6     |           15.2           |        13.3       |              11.9           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((16, 2048, 7168))    |     27.9     |           27.6           |        19.3       |              18.4           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((32, 2048, 7168))    |     43.0     |           42.7           |        28.1       |              27.3           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((64, 2048, 7168))    |     57.8     |           57.1           |        36.4       |              35.9           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((128, 2048, 7168))   |     66.3     |           65.9           |        41.0       |              41.3           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((256, 2048, 7168))   |     70.7     |           70.1           |        45.0       |              44.7           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((512, 2048, 7168))   |     78.5     |           77.8           |        51.0       |              50.8           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((1024, 2048, 7168))  |     89.8     |           89.2           |        64.2       |              63.3           
      nvidia/DeepSeek-R1-FP4, num_experts=256, topk=8, per_act_token=False per_out_ch=False, MKN=((2048, 2048, 7168))  |    133.1     |          132.5           |        91.0       |              90.5           

```
Latest perf: with torch::empty inside the cuda graph replay.
